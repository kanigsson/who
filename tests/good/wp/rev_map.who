type unit
logic tt  : unit
logic /\  : prop -> prop -> prop
logic \/  : prop -> prop -> prop
logic ->  : prop -> prop -> prop
logic mk_2tuple ['a 'b||] : a -> b -> a * b
logic get_2_1_tuple ['a 'b||] : (a * b) -> a
logic get_2_2_tuple ['a 'b||] : (a * b) -> b
logic combine [||'e 'e1 'e2] : < 'e 'e1> -> < 'e1 'e2> -> < 'e 'e1 'e2>
logic restrict [||'e 'e1] : < 'e 'e1> -> < 'e1>
logic empty  : < >
type list ['a] = 
| Nil
| Cons of 'a * 'a list

let rev_map ['a 'b||'e] (f:a ->{  +  'e} b) (l:list[a]) =
  {}
  let rec aux (acc:list[b]) (l1:list[a]) : 'b list,{'e} =
    {}
    match l1 with
      Nil -> acc
    | Cons(x,xs) ->
       let anf = f x in aux (Cons anf acc) xs
    end 
  {}
  in
  aux Nil l
  {}
