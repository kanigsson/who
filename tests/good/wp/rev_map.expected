type unit
logic tt  : unit
logic /\  : prop -> prop -> prop
logic \/  : prop -> prop -> prop
logic ->  : prop -> prop -> prop
logic mk_2tuple ['a 'b||] : a -> b -> a * b
logic get_2_1_tuple ['a 'b||] : (a * b) -> a
logic get_2_2_tuple ['a 'b||] : (a * b) -> b
logic combine [||'e 'e1 'e2] : < 'e 'e1> -> < 'e1 'e2> -> < 'e 'e1 'e2>
logic restrict [||'e 'e1] : < 'e 'e1> -> < 'e1>
logic empty  : < >
type list ['a||] = | Nil | Cons of a* list[a] 
let rev_map ['a
  'b||'e] = (fun (f:(a -> < 'e> -> prop) * (a -> < 'e> -> < 'e> -> b ->
            prop)) ->
            mk_2tuple [list[a] -> < 'e> -> prop,list[a] -> < 'e> -> < 'e> ->
            list[b] -> prop||] (fun (l:list[a]) -> (fun (cur:< 'e>) -> True))
            (fun (l:list[a]) ->
            (fun (old:< 'e>) ->
            (fun (cur:< 'e>) -> (fun (anon:list[b]) -> True))))) 
goal rev_map_correct: forall ['a
  'b||'e]. forall (f:(a -> < 'e> -> prop) * (a -> < 'e> -> < 'e> -> b ->
           prop)).
           forall (l:list[a]).
           forall (s:< 'e>).
           let
           aux =
                (fun (acc:list[b]) ->
                mk_2tuple [list[a] -> < 'e> -> prop,list[a] -> < 'e> ->
                < 'e> -> list[b] -> prop||]
                (fun (l1:list[a]) -> (fun (cur:< 'e>) -> True))
                (fun (l1:list[a]) ->
                (fun (old:< 'e>) ->
                (fun (cur:< 'e>) -> (fun (anon:list[b]) -> True)))))
           in
           (forall (acc:list[b]).
            forall (l1:list[a]).
            forall (s1:< 'e>).
            match l1 with Nil ->  True  | Cons(x,xs) ->
                           (get_2_1_tuple [a -> < 'e> -> prop,a -> < 'e> ->
                            < 'e> -> b -> prop||] f x s1)
                           /\
                           (forall (s2:< 'e>).
                            forall (anon:b).
                            (get_2_2_tuple [a -> < 'e> -> prop,a -> < 'e> ->
                             < 'e> -> b -> prop||] f x s1 s2 anon)
                            ->
                            (get_2_1_tuple [list[a] -> < 'e> ->
                             prop,list[a] -> < 'e> -> < 'e> -> list[b] ->
                             prop||] (aux (Cons [b||] anon acc)) xs s2))  end )
           /\
           (get_2_1_tuple [list[a] -> < 'e> -> prop,list[a] -> < 'e> ->
            < 'e> -> list[b] -> prop||] (aux Nil [b||]) l s)
